File: backend/services/ocr_service.py
import os
import cv2
import json
import numpy as np
from typing import List, Tuple, Dict, Any
from PIL import Image
from dotenv import load_dotenv
from rapidocr_onnxruntime import RapidOCR

load_dotenv()

class OCRService:
    def __init__(self, gemini_api_key: str = None):
        # Initialize RapidOCR local engine as high-speed reliable fallback
        self.local_engine = RapidOCR()
        
        self.api_key = gemini_api_key or os.getenv("GEMINI_API_KEY")
        self.client = None
        if self.api_key:
            try:
                from google import genai
                self.client = genai.Client(api_key=self.api_key)
                print("[INFO] Gemini Vision client initialized for Bounding Box generation.")
            except Exception as e:
                print(f"[WARN] Gemini client init error: {e}")

    def run_ocr(self, image_path: str) -> Tuple[List[Dict[str, Any]], float]:
        """
        Executes text detection and recognition on target image path.
        Prioritizes Gemini Vision model for intelligent semantic bounding boxes.
        Falls back seamlessly to local RapidOCR if Gemini is offline, throttled, or unconfigured.
        """
        if not os.path.exists(image_path):
            raise FileNotFoundError(f"Image not found at {image_path}")

        # Try Gemini Vision model for bounding boxes if client is configured
        if self.client:
            try:
                detections, avg_conf = self._run_gemini_vision_bboxes(image_path)
                if detections:
                    print(f"[INFO] Generated {len(detections)} bounding boxes via Gemini Vision.")
                    return detections, avg_conf
            except Exception as e:
                print(f"[WARN] Gemini Vision bounding box generation failed: {e}. Using local RapidOCR fallback.")

        # Fallback to local RapidOCR
        return self._run_local_ocr(image_path)

    def _run_gemini_vision_bboxes(self, image_path: str) -> Tuple[List[Dict[str, Any]], float]:
        image = Image.open(image_path)
        orig_w, orig_h = image.size

        prompt = """
Locate all printed text and statutory declarations on this Indian packaged commodity.
Detect each label element and its 2D bounding box.

Return a valid JSON array of objects with:
- "text": string (the exact printed text detected)
- "box_2d": [ymin, xmin, ymax, xmax] (normalized integers from 0 to 1000)
- "confidence": float between 0.0 and 1.0

Example:
[
  {"text": "TATA Salt", "box_2d": [343, 267, 432, 696], "confidence": 0.99}
]
"""
        response = self.client.models.generate_content(
            model='gemini-flash-lite-latest',
            contents=[image, prompt],
            config={'response_mime_type': 'application/json'}
        )

        data = json.loads(response.text)
        if not isinstance(data, list):
            return [], 0.0

        detections = []
        confidences = []

        for item in data:
            box = item.get("box_2d")
            text = str(item.get("text", "")).strip()
            conf = float(item.get("confidence", 0.95))

            if box and len(box) == 4 and text:
                ymin, xmin, ymax, xmax = box
                # Scale from 0-1000 normalized to pixel coordinates matching image aspect
                # Scan.tsx SVG canvas scales to 800x500
                canvas_w = 800.0
                canvas_h = 500.0
                
                detections.append({
                    "x_min": round((xmin / 1000.0) * canvas_w, 2),
                    "y_min": round((ymin / 1000.0) * canvas_h, 2),
                    "x_max": round((xmax / 1000.0) * canvas_w, 2),
                    "y_max": round((ymax / 1000.0) * canvas_h, 2),
                    "confidence": conf,
                    "text": text
                })
                confidences.append(conf)

        avg_conf = float(np.mean(confidences)) if confidences else 0.0
        return detections, avg_conf

    def _run_local_ocr(self, image_path: str) -> Tuple[List[Dict[str, Any]], float]:
        result, elapse = self.local_engine(image_path)
        if not result:
            return [], 0.0

        detections = []
        confidences = []
        for item in result:
            bbox, text, conf = item
            pts = np.array(bbox)
            x_min = float(np.min(pts[:, 0]))
            y_min = float(np.min(pts[:, 1]))
            x_max = float(np.max(pts[:, 0]))
            y_max = float(np.max(pts[:, 1]))

            detections.append({
                "x_min": x_min,
                "y_min": y_min,
                "x_max": x_max,
                "y_max": y_max,
                "confidence": float(conf),
                "text": str(text).strip()
            })
            confidences.append(float(conf))

        avg_conf = float(np.mean(confidences)) if confidences else 0.0
        return detections, avg_conf

File: server/src/services/ocrService.js
import fs from 'fs';
import path from 'path';
import { createWorker } from 'tesseract.js';
import dotenv from 'dotenv';

dotenv.config();

export class OCRService {
  /**
   * Run OCR on image file
   * @param {string} imagePath 
   * @returns {Promise<{detections: Array<{text: string, confidence: number, x_min: number, y_min: number, x_max: number, y_max: number}>, avg_conf: number}>}
   */
  static async runOCR(imagePath) {
    if (!fs.existsSync(imagePath)) {
      throw new Error(`Image not found at ${imagePath}`);
    }

    // 1. Try Gemini Vision if API key is provided
    const geminiApiKey = process.env.GEMINI_API_KEY;
    if (geminiApiKey) {
      try {
        const geminiResult = await this.runGeminiVisionOCR(imagePath, geminiApiKey);
        if (geminiResult && geminiResult.detections.length > 0) {
          return geminiResult;
        }
      } catch (err) {
        console.warn(`[OCRService] Gemini Vision failed (${err.message}). Falling back to Tesseract OCR.`);
      }
    }

    // 2. Fallback to Tesseract OCR
    return await this.runTesseractOCR(imagePath);
  }

  /**
   * Tesseract OCR extraction with word/line bounding boxes
   */
  static async runTesseractOCR(imagePath) {
    let worker = null;
    try {
      worker = await createWorker('eng');
      const ret = await worker.recognize(imagePath);
      const detections = [];
      const confidences = [];

      const lines = ret.data.lines || [];
      const canvasW = 800.0;
      const canvasH = 500.0;
      const imgW = ret.data.image_width || 800;
      const imgH = ret.data.image_height || 500;

      for (const line of lines) {
        const text = line.text.trim();
        if (text) {
          const conf = (line.confidence || 85) / 100.0;
          const bbox = line.bbox || { x0: 0, y0: 0, x1: 100, y1: 20 };
          
          detections.push({
            text: text,
            confidence: conf,
            x_min: Math.round((bbox.x0 / imgW) * canvasW),
            y_min: Math.round((bbox.y0 / imgH) * canvasH),
            x_max: Math.round((bbox.x1 / imgW) * canvasW),
            y_max: Math.round((bbox.y1 / imgH) * canvasH)
          });
          confidences.push(conf);
        }
      }

      await worker.terminate();

      const avgConf = confidences.length > 0 
        ? confidences.reduce((a, b) => a + b, 0) / confidences.length 
        : 0.85;

      return {
        detections,
        avg_conf: Math.round(avgConf * 100) / 100
      };
    } catch (error) {
      if (worker) {
        try { await worker.terminate(); } catch (e) {}
      }
      console.error(`[OCRService] Tesseract error: ${error.message}`);
      // Return basic fallback
      return { detections: [], avg_conf: 0.0 };
    }
  }

  /**
   * Gemini Vision model for intelligent 2D bounding boxes and statutory extraction
   */
  static async runGeminiVisionOCR(imagePath, apiKey) {
    const imageBuffer = fs.readFileSync(imagePath);
    const base64Image = imageBuffer.toString('base64');
    const ext = path.extname(imagePath).toLowerCase().replace('.', '') || 'jpeg';
    const mimeType = ext === 'png' ? 'image/png' : 'image/jpeg';

    const prompt = `Locate all printed text and statutory declarations on this Indian packaged commodity.
Detect each label element and its 2D bounding box.
Return a valid JSON array of objects with:
- "text": string (the exact printed text detected)
- "box_2d": [ymin, xmin, ymax, xmax] (normalized integers from 0 to 1000)
- "confidence": float between 0.0 and 1.0

Example:
[
  {"text": "MRP Rs. 50.00 (Incl. of all taxes)", "box_2d": [343, 267, 432, 696], "confidence": 0.99}
]`;

    const url = `https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-flash:generateContent?key=${apiKey}`;
    const response = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        contents: [
          {
            parts: [
              { text: prompt },
              {
                inline_data: {
                  mime_type: mimeType,
                  data: base64Image
                }
              }
            ]
          }
        ],
        generationConfig: {
          response_mime_type: 'application/json'
        }
      })
    });

    if (!response.ok) {
      throw new Error(`Gemini API error ${response.status}: ${await response.text()}`);
    }

    const resData = await response.json();
    const candidateText = resData.candidates?.[0]?.content?.parts?.[0]?.text;
    if (!candidateText) {
      return { detections: [], avg_conf: 0.0 };
    }

    const parsedArray = JSON.parse(candidateText);
    if (!Array.isArray(parsedArray)) {
      return { detections: [], avg_conf: 0.0 };
    }

    const detections = [];
    const confidences = [];
    const canvasW = 800.0;
    const canvasH = 500.0;

    for (const item of parsedArray) {
      const box = item.box_2d;
      const text = (item.text || '').trim();
      const conf = parseFloat(item.confidence || 0.95);

      if (box && box.length === 4 && text) {
        const [ymin, xmin, ymax, xmax] = box;
        detections.push({
          text,
          confidence: conf,
          x_min: Math.round((xmin / 1000.0) * canvasW),
          y_min: Math.round((ymin / 1000.0) * canvasH),
          x_max: Math.round((xmax / 1000.0) * canvasW),
          y_max: Math.round((ymax / 1000.0) * canvasH)
        });
        confidences.push(conf);
      }
    }

    const avgConf = confidences.length > 0
      ? confidences.reduce((a, b) => a + b, 0) / confidences.length
      : 0.95;

    return {
      detections,
      avg_conf: Math.round(avgConf * 100) / 100
    };
  }
}
